pkg/certificate/providers/mrcclient.go

```diff
@@ -0,0 +1,77 @@
+package providers
+
+import (
+	"context"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/k8s/informers"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	errNilInformerCollection = errors.New("informer collection is nil")
+)
+
+// TODO: Rename (keithmattix)
+type MRCClientImpl struct {
+	informerCollection *informers.InformerCollection
+	MRCProviderGenerator
+}
+
+func (m *MRCClientImpl) List() ([]*v1alpha2.MeshRootCertificate, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+	// informers return slice of pointers so we'll convert them to value types before returning
+	mrcPtrs := m.informerCollection.List(informers.InformerKeyMeshRootCertificate)
+	var mrcs []*v1alpha2.MeshRootCertificate
+	for _, mrcPtr := range mrcPtrs {
+		if mrcPtr == nil {
+			continue
+		}
+		mrc, ok := mrcPtr.(*v1alpha2.MeshRootCertificate)
+		if !ok {
+			continue
+		}
+		mrcs = append(mrcs, mrc)
+	}
+
+	return mrcs, nil
+}
+
+func (m *MRCClientImpl) Watch(ctx context.Context) (<-chan certificate.MRCEvent, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+
+	eventChan := make(chan certificate.MRCEvent)
+	m.informerCollection.AddEventHandler(informers.InformerKeyMeshRootCertificate, cache.ResourceEventHandlerFuncs{
```

Comment:
Not sure on the exact functionality here, but if there are multiple event handlers for the same type are they called in a specific order? For example, if there are multiple addFuncs defined for the MRC in what order are they invoked? Just curious :)

Reply:
Good question! So this is what the Go documentation for that function says (emphasis mine):
> AddEventHandler adds an event handler to the shared informer using the shared informer's resync period.  Events to a single handler are delivered sequentially, **but there is no coordination between different handlers.**

So there's no coordination between subscribers to these events, but events to a single subscriber are guaranteed to be delivered in-order.

Reply:
To clarify: Are they guaranteed to be in order from the k8s master? ie: if a user does a delete then an update on an object is that guaranteed to be in order?

I'd guess that k8s doesn't place those guarantees. I'd expect that missed updates are also possible (this is probably why they have a resync interval exposed)

Reply:
They guarantee that received events will be in order, NOT that every state will be enumerated. From the sharedinformer docs:
> A SharedInformer maintains a local cache...of he state of each relevant object.  This cache is eventually consistent with the authoritative state.  This means that, unless prevented by persistent communication problems, if ever a particular object ID X is authoritatively associated with a state S then for every SharedInformer I whose collection includes (X, S) eventually either (1) **I's cache associates X with S or a later state of X**, (2) I is stopped, or (3) the authoritative state service for X terminates.

So we won't get out-of-order updates, but updates may be coalesced

Reply:
Oh that's a fantastic point. I don't think we have to worry about that at this layer, but subscribers should be aware.